docs/ExtensibleMaps.md

```diff
@@ -0,0 +1,133 @@
+program type specific, and are implemented / coordinated by the extensions.
+
+Extensible / custom maps are program type specific map types that can be implemented / coordinated by eBPF extensions
+(e.g. BPF_MAP_TYPE_XSKMAP) through a new NMR (Network Module Registrar) provider interface. This document contains
+the design for adding support for custom maps in eBPF-for-Windows. Custom maps enable extensions to register and
+manage their own map types beyond those provided by the core eBPF runtime.
+
+Custom maps will be based on one of the underlying / base map types already implemented in eBPFCore. When extensions
+declare a new custom map type, they also declare the base map type on which the custom map should be based on.
+For example, BPF_MAP_TYPE_XSKMAP can be based on the existing BPF_MAP_TYPE_HASH_MAP.
+With this approach, eBPFCore can implement the custom map, and extensions will provide hooks / callbacks for map
+creation, and CRUD operations on the map.
+
+This approach has a benefit that extensions do not need to re-implement a performant, RCU-aware data structure and
+can leverage the implementation that is present in eBPFCore.
+
+Currently, eBPF-for-Windows will only allow base map type BPF_MAP_TYPE_HASH_MAP. This support can be extended to other
+base map types in future based on the requirements.
+
+Note: If there is a need for an extension to implement a map type that cannot be based on any existing map type, we can
+extend this interface for extensions to optionally provide their whole implementation, instead of relying on a base map
+type in eBPF-for-Windows.
+
+## NMR Interface for Extensions
+To implement custom maps, a new NMR interface will be added and extensions will register as map info providers using
+NMR, similar to existing program and hook providers.
+
+**New NMR Provider Interface**: `EBPF_MAP_INFO_EXTENSION_IID`
+
+**Provider Registration Data**:
+```c
+typedef struct _ebpf_map_provider_data {
+    ebpf_extension_header_t header;
+    uint32_t custom_map_type;  // Single map type per provider
+    uint32_t base_map_type;
+    ebpf_map_provider_dispatch_table_t* dispatch_table;
+} ebpf_map_provider_data_t;
+```
+
+**Provider Dispatch Table**:
+```c
+typedef struct _ebpf_map_provider_dispatch_table {
+    ebpf_extension_header_t header;
+    ebpf_map_create_notify_t create_map_notify;
+    ebpf_map_delete_notify_t delete_map_notify;
+    ebpf_map_associate_program_type_t associate_program;
+    ebpf_map_find_element_notify_t find_element_notify;
+    ebpf_map_update_element_notify_t update_element_notify;
+    ebpf_map_delete_element_notify_t delete_element_notify;
+    ebpf_map_get_next_key_and_value_notify_t get_next_key_and_value_notify;
+} ebpf_map_provider_dispatch_table_t;
+```
+
+An extension (provider) needs to provide the above dispatch table. eBPF runtime will invoke the above functions in the
```

Comment:
Must all fields be non-null?  Or can null be passed for an operation that isn't supported?  Or null to mean use the base map type's implementation?